### PR TITLE
ENH: offset parameter for MultipleLocator

### PR DESCRIPTION
**Pull request by rcomer**
## PR Summary

Picked up from #21168.  Adds an *offset* parameter to `MultipleLocator` so the ticks can all be shifted by a given value.

```python
import matplotlib.pyplot as plt
import matplotlib.ticker as mticker

_, ax = plt.subplots()
ax.plot(range(10))
locator = mticker.MultipleLocator(base=3, offset=0.3)
ax.xaxis.set_major_locator(locator)

plt.show()
```

test

~### To Do~

~1. Agree whether *offset* is the best name for the new parameter.~
~2. Address documentation tasks in the below checklist, but I wanted to get feedback on the principle and the parameter name before I went all in.~

## PR Checklist

**Documentation and Tests**
- [X] Has pytest style unit tests (and `pytest` passes)
- [x] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] New plotting related features are documented with examples.

**Release Notes**
- [x] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [N/A] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [x] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

### CODE REVIEW

**Comment by rcomer** [comment]
I pushed a change without rebasing just to see what Mypy would do.  It's running....

**Comment by rcomer** [comment]
@ksunden could you help me decipher this Mypy failure?

File: galleries/examples/ticks/tick-locators.py [hunk 1]
```diff
@@ -53,7 +53,7 @@ def setup(ax, title):
 
 # Index Locator
 setup(axs[4], title="IndexLocator(base=0.5, offset=0.25)")
-axs[4].plot(range(0, 5), [0]*5, color='white')
+axs[4].plot([0]*5, color='white')
```
Comment (rcomer) [inline galleries/examples/ticks/tick-locators.py:56]: It's not completely obvious from looking at this example what the difference is between `MultipleLocator` and `IndexLocator`.  But from the [API docs](https://matplotlib.org/devdocs/api/ticker_api.html#matplotlib.ticker.IndexLocator) it seems to me that `IndexLocator` makes most sense when the axis does not have numbers associated with it.  So I took out the x-values.

File: ci/mypy-stubtest-allowlist.txt [hunk 2]
```diff
@@ -60,7 +60,6 @@ matplotlib.axis.Axis.draw
 matplotlib.offsetbox.*Offset[Bb]ox.get_offset
 
 # Inconsistent super/sub class signatures (other than just arg name)
-matplotlib.ticker.MultipleLocator.set_params
```
Comment (rcomer) [inline ci/mypy-stubtest-allowlist.txt:63]: My understanding of this is:

* this file exists to suppress errors that Mypy would throw but we've decided we don't care about
* Before this PR `set_params` didn't have a default value for _base_, but the stub suggested it should, so without this suppression line we would have got
  ```
   error: matplotlib.ticker.MultipleLocator.set_params is inconsistent, stub argument "base" has a default value but runtime argument does not
   ```
* Now that I have given _base_ a default value, Mypy detected that this line was not needed so gave me a different error
   ```
   note: unused allowlist entry matplotlib.ticker.MultipleLocator.set_params
   ```
* From the comment on line 62, I assume the stub generation depends on the parent class somehow (which just has `set_params(self, **kwargs)`.  But I have not understood this bit.

**Comment by rcomer** [comment]
I think this is now ready for review.  I have not [yet] added a what's new entry, as I'm still pretty unclear in general what the criteria are for deciding to do so.

**Comment by oscargus** [comment]
> I'm still pretty unclear in general what the criteria are for deciding to do so.

``.. versionadded:: 3.8`` I think  (or just that it is nice to showcase some new features!).

(I happened to read https://matplotlib.org/stable/devel/coding_guide.html#new-features-and-api-changes the other day...)

Parameter makes sense! (Will review code when I have a bit more time.)

**Comment by jklymak** [comment]
Yes this should definitely get what's new entry

**Comment by rcomer** [comment]
OK thanks.  What's new now added.

File: ci/mypy-stubtest-allowlist.txt (hunk 2 above)
Comment (rcomer) [reply ci/mypy-stubtest-allowlist.txt]: This line was removed at #25844, so this comment is now irrelevant.

**Review by QuLogic — APPROVED** [review]

**Review by ksunden — APPROVED** [review]

**Pull request merged**